kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogReplaySuite.scala

```diff
@@ -13,84 +13,264 @@
-        }
+  test("simple end to end with inserts and deletes and checkpoint") {
+    val expectedValues = (0L until 5L) ++ (10L until 15L) ++ (20L until 25L) ++
+      (30L until 35L) ++ (40L until 45L) ++ (50L to 65L)
+    checkTable(
+      path = goldenTablePath("basic-with-inserts-deletes-checkpoint"),
+      expectedAnswer = expectedValues.map(TestRow(_)),
+      expectedSchema = new StructType().add("id", LongType.LONG),
+      expectedVersion = Some(13L)
+    )
+  }
+
+  test("simple end to end with inserts and updates") {
+    val expectedValues = (0 until 50).map((_, "N/A")) ++
+      (50 until 100).map(x => (x, s"val=$x"))
+    checkTable(
+      path = goldenTablePath("basic-with-inserts-updates"),
+      expectedAnswer = expectedValues.map(TestRow.fromTuple)
+    )
+  }
+
+  test("simple end to end with inserts and merge") {
+    val expectedValues = (10 until 50).map(x => (x, s"val=$x")) ++
+      (50 until 100).map((_, "N/A")) ++ (100 until 150).map((_, "EXT"))
+    checkTable(
+      path = goldenTablePath("basic-with-inserts-merge"),
+      expectedAnswer = expectedValues.map(TestRow.fromTuple)
+    )
+  }
+
+  test("simple end to end with restore") {
+    checkTable(
+      path = goldenTablePath("basic-with-inserts-overwrite-restore"),
+      expectedAnswer = (0L until 200L).map(TestRow(_)),
+      expectedVersion = Some(3)
+    )
+  }
+
+  test("end to end only checkpoint files") {
+    val expectedValues = (5L until 10L) ++ (0L until 20L)
+    checkTable(
+      path = goldenTablePath("only-checkpoint-files"),
+      expectedAnswer = expectedValues.map(TestRow(_))
+    )
+  }
+
+  Seq("protocol", "metadata").foreach { action =>
+    test(s"missing $action should fail") {
+      val path = goldenTablePath(s"deltalog-state-reconstruction-without-$action")
+      val e = intercept[IllegalStateException] {
+        latestSnapshot(path).getSchema(defaultTableClient)
       }
-    } finally scanFilesBatchIter.close()
-    dataBatches
+      assert(e.getMessage.contains(s"No $action found"))
+    }
   }
 
-  test("simple end to end with inserts and deletes and checkpoint") {
-    val unresolvedPath = GoldenTableUtils.goldenTablePath("basic-with-inserts-deletes-checkpoint")
-    val table = io.delta.kernel.Table.forPath(tableClient, unresolvedPath)
-    val conf = new Configuration()
-    val client = DefaultTableClient.create(conf)
-    val snapshot = table.getLatestSnapshot(client)
-    val schema = snapshot.getSchema(client)
-    val version = snapshot.getVersion(client)
-
-    assert(schema === new StructType().add("id", LongType.LONG))
-    assert(version === 13)
-
-    var output = new scala.collection.mutable.ArrayBuffer[Long]()
-    readSnapshot(schema, snapshot).foreach { data =>
-      val rowsIter = data.getRows
-      while (rowsIter.hasNext) {
-        val row = rowsIter.next()
-        output += row.getLong(0)
+  // TODO missing protocol should fail when missing from checkpoint
```

Comment:
Is this a bug in Kernel or just the golden table needs to be updated?

Reply:
Just golden table